[DOCS] Added instructions for 3 nodes + kibana (#586)
elastic/stack-docs: docs/en/getting-started/get-started-docker.asciidoc

> [[get-started-docker]]

Reviewer: I wasn't able to run the non-secured example with the default docker for mac settings (2 CPUs and 2 GiB Memory) without at least one of the nodes failing. I bumped up the allocation to 4 CPU and 4 GiB Memory (Docker for Mac -> Settings -> Advanced) and tested and it appears to work fine. I'm thinking we might want to add a prerequisite section that outlines the resource requirements for running these configurations (while my test worked, it would be good to have someone from the ES team weigh in on optimal recommendations).

Reviewer 2: I agree we should clarify that the user needs to check that there's sufficient memory in Docker for Mac (I suppose also for Windows) and we recommend at least 4GiB -- given that these examples have 512MiB heap (plus half of that for Max Direct Memory) x 3 + enough memory for Kibana. Also we should include the Docker link that it can be changed via https://docs.docker.com/docker-for-mac/#resources

Author: In my own testing, once I bumped to 4 GiB things worked fine. Agree that needs to be called out.

== Running the {stack} on Docker

The Elastic Docker registry contains Docker images for all the products in 
the {stack}: https://www.docker.elastic.co/

[float]
[[run-stack-docker]]
=== Run with Docker Compose

To get the default distributions of {es} and {kib} up and running in Docker, 

> you can use `docker-compose`:

Reviewer 3: > `docker-compose`

Not sure if the literals are needed here. I'd use "Docker Compose" but don't feel strongly.

> . Pull the docker images for {es} and {kib}:

Reviewer 3: > docker images 

Should Docker be capitalized here? "container image" would work fine too.

+
["source","sh",subs="attributes"]
--------------------------------------------

> docker pull docker.elastic.co/elasticsearch/elasticsearch:{version}

Reviewer: We can eliminate the explicit docker pull steps if we want given that the images are specified in the docker-compose files (and the version gets supplied when the command is run). I don't personally see any benefit in explicitly calling out the pull step and it would allow us to reduce the number of documented steps/user actions if we remove.

docker pull docker.elastic.co/kibana/kibana:{version}
--------------------------------------------

. Create a `docker-compose.yml` file for the Elastic Stack:
+
["source","yaml"]
--------------------------------------------
include::docker/docker-compose.yml[]
--------------------------------------------

> . Run `docker-compose` to bring up a three node Elasticsearch cluster and Kibana:

Reviewer 3: ```suggestion
. Run `docker-compose` to bring up a three node {es} cluster and {kib}:
```

+
["source","sh",subs="attributes"]
--------------------------------------------
VERSION={version} docker-compose up
--------------------------------------------

> . Submit a `_cat/nodes` request to see that the nodes are up and running:

Reviewer 3: We may want to link to the cat nodes API here.

It also may be confusing to include a console snippet *before* we explicitly tell people to open up Kibana.

Author: In the getting started content, I'm opting to minimize linking folks out of the flow.

Reviewer 3: Makes sense. Thanks!

+
[source,console]
--------------------------------------------------
GET /_cat/nodes?v
--------------------------------------------------

> . Open Kibana to load sample data and interact with the cluster: 

Reviewer 3: ```suggestion
. Open {kib} to load sample data and interact with the cluster: 
```

http://localhost:5601. 

[float]
[[get-started-docker-tls]]
=== Run in Docker with TLS enabled 

> When security is enabled with a Gold or Platinum license, 

Reviewer 3: > Gold or Platinum license
links?

Reviewer 3: Can discard.

Author: Here I agree that we want to xref the subscriptions page, as they might have questions about the license levels before they dive into the procedure.

> TLS encryption must be configured for the {es} transport layer. 

Reviewer 3: > TLS encryption
It may be good to spell out "Transport Layer Security (TLS)"

and a {kib} instance with TLS enabled.

.`instances.yml`:
["source","yaml"]
----
include::docker/instances.yml[] 
----

.`.env`:
["source","txt",subs="attributes"]
----
include::docker/.env[]
----

.`create-certs.yml`:
["source","txt"]
----
include::docker/create-certs.yml[]
----

.`docker-elastic-tls.yml`:
["source","txt"]
----
include::docker/elastic-docker-tls.yml[]
----
<1> Bootstrap {es} with the password defined in `.env` to enable you to run 
the `elasticsearch-setup-passwords` tool to set the passwords for the built in users, 
including the `elastic` and `kibana` users. 
See {ref}/bootstrap-elastic-passwords.html[Bootstrap Elastic passwords].
<2> Generate and apply a trial license that supports Transport Layer Security.
<3> Enable Transport Layer Security to encrypt client communications.
<4> Enable Transport Layer Security to encrypt internode communications.
<5> Allow the use of self-signed certificates by not requiring hostname verification. 

--

. Generate certificates for {es} by bringing up the `create-certs` container:
+
--
["source","sh"]
----
docker-compose -f create-certs.yml run --rm create_certs
----

--

> . Bring up the three node {es} cluster:

Reviewer 3: ```suggestion
. Bring up the three-node {es} cluster:
```

+
--
["source","sh"]
----
docker-compose -f elastic-docker-tls.yml up -d
----

> IMPORTANT: At this point, Kibana cannot connect to the {es} cluster.

Reviewer 3: ```suggestion
IMPORTANT: At this point, {kib} cannot connect to the {es} cluster.
```

You must generate a password for the built-in `kibana` user, update the `ELASTICSEARCH_PASSWORD` 

> in the compose file, and restart to enable Kibana to communicate with the secured cluster. 

Reviewer 3: ```suggestion
in the compose file, and restart to enable {kib} to communicate with the secured cluster. 
```

--

. Run the `elasticsearch-setup-passwords` tool to generate passwords for all built-in users, 
including the `kibana` user:
+
--
["source","sh"]
----
docker exec es01 /bin/bash -c "bin/elasticsearch-setup-passwords \
auto --batch \
-Expack.security.http.ssl.certificate=certificates/es01/es01.crt \
-Expack.security.http.ssl.certificate_authorities=certificates/ca/ca.crt \
-Expack.security.http.ssl.key=certificates/es01/es01.key \
--url https://localhost:9200"
----

IMPORTANT: If you don't use PowerShell on Windows, remove the trailing `\`characters
and join the lines before running this command.

--

. Set `ELASTICSEARCH_PASSWORD` in the `elastic-docker-tls.yml` compose file to the password 
generated for the `kibana` user.
+
--
["source","txt",subs="-callouts"]
----
include::docker/elastic-docker-tls.yml[]
----

--

> . Use `docker-compose` to restart the cluster and Kibana:

Reviewer 3: ```suggestion
. Use `docker-compose` to restart the cluster and {kib}:
```

+
--
["source","sh"]
----
docker-compose down
docker-compose -f elastic-docker-tls.yml up -d
----
--

> . Open Kibana to load sample data and interact with the cluster: 

Reviewer 3: ```suggestion
. Open {kib} to load sample data and interact with the cluster: 
```

https://localhost:5601. 
+
NOTE: Because SSL is also enabled for communications between {kib} and client browsers, 
you must access {kib} via the HTTPS protocol. 

[float]
[[load-settings-file]]
==== Loading settings from a file

> Specifying settings for {es} and {kibana} directly in the compose file is a convenient way to get started, 

Reviewer 3: This one won't render correctly as-is.
```suggestion
Specifying settings for {es} and {kib} directly in the compose file is a convenient way to get started, 
```

== Running the {stack} on Docker

The Elastic Docker registry contains Docker images for all the products in the 
{stack}: https://www.docker.elastic.co/

To get the default distributions of {es} and {kib} up and running in Docker, you can use 
`docker-compose`:

. Pull the docker images for {es} and {kib}:
+
["source","sh",subs="attributes"]
--------------------------------------------
docker pull docker.elastic.co/elasticsearch/elasticsearch:{version}
docker pull docker.elastic.co/kibana/kibana:{version}
--------------------------------------------

. {stack-gs}/docker-compose.yml[Download the docker-compose.yml] file for the Elastic Stack:
+
["source","yaml",subs="attributes"]
--------------------------------------------
include::images/docker-compose.yml[]
--------------------------------------------

. Run `docker-compose` to bring up a three node Elasticsearch cluster and Kibana:
+
["source","sh",subs="attributes"]
--------------------------------------------

> docker-compose up

Reviewer 2: Does it make sense to document the link (http://localhost:5601) to visit Kibana and continue from there?
